### PR TITLE
Add test case for uefi secureboot

### PR DESCRIPTION
**Pull request by vivianQizhu**
id: 1692691
Signed-off-by: qizhu <qizhu@redhat.com>

### CODE REVIEW

**Comment by vivianQizhu** [comment]
Test passed on both RHEL and Windows.
(1/1) Host_RHEL.m8.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: PASS (2352.83 s)
 (1/2) Host_RHEL.m8.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: PASS (161.81 s)
 (2/2) Host_RHEL.m8.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.x.0.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: PASS (694.14 s)

**Comment by vivianQizhu** [comment]
@yanan-fu Please help review. Thanks.

**Comment by vivianQizhu** [comment]
@ChongkaiMa

**Comment by ChongkaiMa** [comment]
@vivianQizhu  Hi qianqian, I have tested this case in RHEL7.7 host. There were some thing wrong when I used RHEL7.6 guest and RHEL7.7 guest. RHEL8.0.1 guest and Win2019 guest was fine. Please have a look. 

(001/200) repeat1.Host_RHEL.m7.u7.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: ERROR: Install pesign failed, please install it before test. (1454.02 s)
 (002/200) repeat1.Host_RHEL.m7.u7.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: ERROR: Install pesign failed, please install it before test. (1529.19 s)
 (003/200) repeat1.Host_RHEL.m7.u7.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: PASS (1345.56 s)
 (004/200) repeat1.Host_RHEL.m7.u7.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: PASS (2210.90 s)

**Comment by ChongkaiMa** [comment]
@vivianQizhu Continue from last comment. I pretty sure I have installed package pesign in the host. I can do that by manual. 

#pesign --show-signature -i vmlinuz
---------------------------------------------
certificate address is 0x7f06e6e31728
Content was not encrypted.
Content is detached; signature cannot be verified.
The signer's common name is Red Hat Secure Boot Signing 3 (beta)
No signer email address.
Signing time: Mon May 06, 2019
There were certs or crls included.
---------------------------------------------

**Comment by vivianQizhu** [comment]
@ChongkaiMa Hi, we are checking pesign for guest, so if you install it in your `host`, it won't work. The error `ERROR: Install pesign failed, please install it before test. ` is because RHEL7 has not pesign in its repo by default. So if you have any clue about how to install it on RHEL7 os, I would add the installation method to the code. Otherwise, we could choose RHEL8 for our test, anyway I believe it's OS independent so there is no difference.

**Comment by ChongkaiMa** [comment]
@vivianQizhu Hi, thanks qianqian, I fully understand. There are 2 thing I am thinking about.
 
[1] Do we need to update the ERROR information? If follow the information, I will try to find some where can install pesign first, then start the test. But pesign is installed inside the guest. How can I install it before the test? I do not have guest before the test.

[2] In my understanding, we can not test RHEL7 guest until I find some where can install pesign, however; from you comment in Bug 1692691#c11, it is not suitable for other upstream users if down load the package from the brewweb. Therefore, RHEL7 guest can not be tested by auto. Am I right? 

In my opinion, RHEL7 guest should be tested by auto. We need to find some thing way out.

**Comment by vivianQizhu** [comment]
> @vivianQizhu Hi, thanks qianqian, I fully understand. There are 2 thing I am thinking about.
> 
> [1] Do we need to update the ERROR information? If follow the information, I will try to find some where can install pesign first, then start the test. But pesign is installed inside the guest. How can I install it before the test? I do not have guest before the test.
> 
Right, it can't be installed before test start, so I should update the error message. 
> [2] In my understanding, we can not test RHEL7 guest until I find some where can install pesign, however; from you comment in Bug 1692691#c11, it is not suitable for other upstream users if down load the package from the brewweb. Therefore, RHEL7 guest can not be tested by auto. Am I right?
> 
> In my opinion, RHEL7 guest should be tested by auto. We need to find some thing way out.

If there is rpm package and no dependency of it, we might put it at /deps directory in tp-qemu, and copy to guest after os installation. I just previously thought it is not necessary since we are testing firmware instead of guest os, so any unsigned and signed guest os is fine.

**Comment by vivianQizhu** [comment]
@ChongkaiMa I will modify the error message and move the discussion about repo to bz. Thanks.

**Comment by vivianQizhu** [comment]
@ChongkaiMa @yanan-fu Updated, please help review, thanks.

**Comment by ChongkaiMa** [comment]
@vivianQizhu I have executed this case on the rhel7.7 host. Windows guest fail. 
FAIL: Secure boot is not enabled

**Comment by vivianQizhu** [comment]
> @vivianQizhu I have executed this case on the rhel7.7 host. Windows guest fail.
> FAIL: Secure boot is not enabled

That means `Confirm-SecureBootUEFI` command failed, would you check your log for the command result? Thanks.

**Comment by vivianQizhu** [comment]
@ChongkaiMa  Thanks for testing, updated and test passed, please help check. @yanan-fu Please help review, thanks.
(1/1) Host_RHEL.m7.u7.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.uefi_secureboot.extra_cdrom_ks.with_installation.q35: /^PASS (2669.15 s)

File: qemu/tests/cfg/uefi_secureboot.cfg [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            cd_format_cd1 = ide
+            cd_format_winutils = ide
+            cd_format_unattended = ide
+        q35:
+            cd_format_cd1 = ahci
+            cd_format_winutils = ahci
+            cd_format_unattended = ahci
+    # Below variants is to share configurations related to installation defined in other cfg
+    variants:
+        - with_installation:
```
Comment (yanan-fu) [inline qemu/tests/cfg/uefi_secureboot.cfg]: Better use `@`

Comment (vivianQizhu) [reply qemu/tests/cfg/uefi_secureboot.cfg]: @yanan-fu Updated, thanks.

File: qemu/tests/uefi_secureboot.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Verify file OVMF_VARS.secboot.fd, boot released OS with Secure boot by default.
+
+    1. Install guest by OVMF environment(With OVMF_VARS.fd)
+    2. Boot guest and check if the guest is signed
+    3. Reboot the guest using file OVMF_VARS.secboot.fd
```
Comment (yanan-fu) [inline qemu/tests/uefi_secureboot.py]: Better use `shutdown and boot the guest with ***`,  `reboot` is ambiguous here.

File: qemu/tests/uefi_secureboot.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            s, o = session.cmd_status_output(install_cmd)
+            if s != 0:
+                test.cancel('Install pesign failed with output: "%s". '
+                            'Please define proper source for guest' % o)
+        vmlinuz = '/boot/vmlinuz-%s' % session.cmd_output('uname -r')
+        check_sign_cmd = check_sign_cmd % vmlinuz
+    sign_info = session.cmd_output(check_sign_cmd)
+    signed = _check_signed()
+    error_context.context('Guest signed status is %s, reboot guest'
+                          ' with secure boot' % signed, logging.info)
```
Comment (yanan-fu) [inline qemu/tests/uefi_secureboot.py]: same with above, about `reboot`.

Comment (vivianQizhu) [reply qemu/tests/uefi_secureboot.py]: @yanan-fu Updated, thanks.

**Comment by ChongkaiMa** [comment]
@yanan-fu @vivianQizhu Hi yfu and qizhu, I have reviewed the code. It looks good. Ack for this code. Thanks

**Comment by vivianQizhu** [comment]
Thanks @ChongkaiMa @yanan-fu

**Pull request merged**